pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCornerCaseTest.java

```diff
@@ -0,0 +1,356 @@
+     *   3: Less than the limitation when building batched messages payload.
+     *   2: Equals the limitation when building batched messages payload.
+     */
+    @DataProvider(name = "maxMessageSizes")
+    public Object[][] maxMessageSizes(){
+        return new Object[][] {
+                {1},
+                {3},
+                {26}
+        };
+    }
+
+    @Test(dataProvider = "maxMessageSizes")
+    public void testBatchedSendMessageSizeExceeded(int maxMessageSize) throws Exception {
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        admin.topics().createNonPartitionedTopic(topicName);
+        ProducerImpl<String> producer = (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING).topic(topicName)
+                .enableBatching(true)
+                .compressionType(CompressionType.NONE)
+                .create();
+        final ClientCnx cnx = producer.getClientCnx();
+        producer.getConnectionHandler().maxMessageSize = maxMessageSize;
+        MsgPayloadTouchableMessageBuilder<String> msgBuilder1 = newMessage(producer);
+        MsgPayloadTouchableMessageBuilder<String> msgBuilder2 = newMessage(producer);
+        /**
+         * Mock an error: reached max message size. see more detail {@link #maxMessageSizes()}.
+         */
+        msgBuilder1.value("msg-1").sendAsync();
+        msgBuilder2.value("msg-1").sendAsync().exceptionally(ex -> {
+            log.warn("reached maxMessageSize", ex);
+            return null;
+        }).join();
+
+        // Verify: message payload has been released.
+        // Since "MsgPayloadTouchableMessageBuilder" has called "buffer.retain" once, "refCnt()" should be "1".
+        producer.close();
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(producer.getPendingQueueSize(), 0);
+        });
+        assertEquals(msgBuilder1.payload.refCnt(), 1);
+        assertEquals(msgBuilder2.payload.refCnt(), 1);
+
+        // cleanup.
+        cnx.ctx().close();
+        msgBuilder1.release();
+        msgBuilder2.release();
+        assertEquals(msgBuilder1.payload.refCnt(), 0);
+        assertEquals(msgBuilder2.payload.refCnt(), 0);
+        admin.topics().delete(topicName);
+    }
+
+    @Test
+    public void testSendAfterClosedProducer() throws Exception {
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        admin.topics().createNonPartitionedTopic(topicName);
+        ProducerImpl<String> producer =
+                (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING).topic(topicName).create();
+        // Publish after the producer was closed.
+        MsgPayloadTouchableMessageBuilder<String> msgBuilder = newMessage(producer);
+        producer.close();
+        msgBuilder.value("msg-1").sendAsync().exceptionally(ex -> {
+            log.warn("expected error", ex);
```

Comment:
If it's an expected error, don't print warn logs. You'd better use `assertThrows` or a trivial `try-catch`:

```java
try {
    msgBuilder.value("msg-1").send();
    fail();
} catch (PulsarClientException ignored) {
    // or you can add the type assertion here
}
```

Reply:
Solved